feat: add interval util functions for priority queue

hi-rustin opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: ref https://github.com/pingcap/tidb/pull/49018

Problem Summary:

### What changed and how does it work?

Added some functions to get the analysis intervals.

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

hi-rustin commented:
/retest

hi-rustin reviewed (commented):
🔢 Self-check

hawkingrei commented on pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:
```diff
@@ -0,0 +1,128 @@
+	partitionNames ...string,
+) (time.Duration, error) {
+	query := ""
+	params := make([]interface{}, 0, len(partitionNames)+3)
```
```suggestion
	params := make([]any, 0, len(partitionNames)+3)
```

hawkingrei commented on pkg/statistics/handle/autoanalyze/priorityqueue/interval_test.go:
```diff
@@ -0,0 +1,274 @@
+)
+
+func TestGetAverageAnalysisDuration(t *testing.T) {
+	store, _ := testkit.CreateMockStoreAndDomain(t)
```
```suggestion
	store := testkit.CreateMockStore(t)
```

hawkingrei commented on pkg/statistics/handle/autoanalyze/priorityqueue/interval_test.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+func TestGetLastFailedAnalysisDuration(t *testing.T) {
+	store, _ := testkit.CreateMockStoreAndDomain(t)
```
```suggestion
	store := testkit.CreateMockStore(t)
```

hi-rustin commented on pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:
```diff
@@ -0,0 +1,128 @@
+	partitionNames ...string,
+) (time.Duration, error) {
+	query := ""
+	params := make([]interface{}, 0, len(partitionNames)+3)
```
```suggestion
	params := make([]any, 0, len(partitionNames)+3)
```

hi-rustin reviewed (commented):
🔢 Self-check

time-and-fate reviewed (commented):
Looks like we didn't consider what would happen if there are no such rows in the analyze history.

time-and-fate commented on pkg/statistics/handle/autoanalyze/priorityqueue/interval.go:
```diff
@@ -0,0 +1,128 @@
+)
+
+const avgDurationQueryForTable = `
+	SELECT AVG(TIMESTAMPDIFF(SECOND, start_time, end_time)) AS avg_duration
+	FROM mysql.analyze_jobs
+	WHERE table_schema = %? AND table_name = %? AND state = 'finished' AND fail_reason IS NULL AND partition_name = ''
+	ORDER BY id DESC
+	LIMIT 5;
```
This SQL is a bit questionable with both an `avg` and `ORDER BY + LIMIT`.

hi-rustin replied:
I want to calculate the average analysis time for the most recent 5 records. Could you please provide more information about the issue that is causing you worry?

time-and-fate replied:
The semantic of a SELECT statement is generally `JOIN -> WHERE -> AGG -> HAVING -> ORDER BY -> LIMIT`. You would need to write it like `avgDurationQueryForPartition` below to get what you want.
You can see this example: https://www.db-fiddle.com/f/3moV1YrnrZchZTaK41fk2h/1

hi-rustin commented:
> Looks like we didn't consider what would happen if there are no such rows in the analyze history.

Then we will return 0 and compare it with 0 by `0 < 0`, then we can analyze it because it doesn't satisfy the condition.

time-and-fate commented:
> > Looks like we didn't consider what would happen if there are no such rows in the analyze history.
> 
> Then we will return 0 and compare it with 0 by `0 < 0`, then we can analyze it because it doesn't satisfy the condition.

Have you tested this case to make sure?
`AVG` and `MIN` on empty result set would be `NULL` instead of `0`.

And I think we'd better add some comments. For example, it's not obvious that `len(rows) == 0` in `getLastFailedAnalysisDuration()` is an expected case.

hi-rustin commented:
> > > Looks like we didn't consider what would happen if there are no such rows in the analyze history.
> > 
> > 
> > Then we will return 0 and compare it with 0 by `0 < 0`, then we can analyze it because it doesn't satisfy the condition.
> 
> Have we tested this case to make sure? `AVG` and `MIN` on empty result set would be `NULL` instead of `0`.
> 
> And I think we'd better add some comments. For example, it's not obvious that `len(rows) == 0` in `getLastFailedAnalysisDuration()` is an expected case.

I added some tests to cover it and also updated the implementation to make it more obvious.

hi-rustin commented:
/retest

hi-rustin commented:
@ti-chi-bot Hi man

hi-rustin commented:
/retest
